### ISSUE TITLE
Cannot determine virtual subclass of ABC defined via __subclasshook__ 

### ISSUE THREAD

**Issue by hwayne** [python/mypy#1809]
```
from abc import ABCMeta

class A(metaclass=ABCMeta):
    @classmethod
    def __subclasshook__(cls, other):
        return True

class B():
    pass

def is_A(a: A) -> bool:
    return isinstance(a, A) # returns True

print(is_A(B())) # error: Argument 1 to "is_A" has incompatible type "B"; expected "A"
```

Similar to #1459, except from`__subclasshook__` instead of `register`. Like #1459 may be something that would require much more complicated changes (like solving the halting problem).

Main use case I can see is that it makes it a lot easier to check duck-typing.

**Comment by hwayne** [comment]
Turns out this is already in the [wiki](https://github.com/python/mypy/wiki/Unsupported-Python-Features). Sorry for the overhead!

**Comment by bugreport1234** [comment]
I'd like to request to reopen this. At the time this bug was reported, the wiki page for unsupported features read "Run-time addition/mutation of base classes (e.g. the abc module)". In the meantime it seems some support for abc was added and now it reads "Run-time addition/mutation of base classes (e.g. registering ABCs)". Subclass checks via `__subclasshook__` seem like a distinct feature from registering ABCs, so from the wording of the wiki page I'd assume using `__subclasshook__` works now. However, the above example still gives the same error, so I assume this is either undocumented or a bug.

The same problem exists for `__subclasscheck__` and `__instancecheck__`.

**Comment by ilevkivskyi** [comment]
From OP
> Main use case I can see is that it makes it a lot easier to check duck-typing.

For duck typing there are [protocols](http://mypy.readthedocs.io/en/latest/class_basics.html#protocols-and-structural-subtyping), use them instead. They even have (limited) support for runtime checks (i.e. automatically adding ``__subclasshook__``) if used with ``@runtime`` decorator.

I don't think we can have reasonable static support for _arbitrary_ ``__subclasshook__`` and friends.

**Comment by maggyero** [comment]
How come that under Mypy, `__subclasshook__` works for one-trick ponies from `collections.abc`, but not for user-defined classes?

For instance, this program

```python
from collections.abc import Hashable

class A:
    def __hash__(self) -> int:
        return 0

a: Hashable = A()
```

outputs

```
$ mypy demo.py --strict
Success: no issues found in 1 source file
```

But this program

```python
from abc import ABCMeta, abstractmethod

def _check_methods(C: type, *methods: str) -> bool:
    mro = C.__mro__
    for method in methods:
        for B in mro:
            if method in B.__dict__:
                if B.__dict__[method] is None:
                    return NotImplemented
                break
        else:
            return NotImplemented
    return True

class Hashable(metaclass=ABCMeta):
    __slots__ = ()

    @abstractmethod
    def __hash__(self) -> int:
        return 0

    @classmethod
    def __subclasshook__(cls, C: type) -> bool:
        if cls is Hashable:
            return _check_methods(C, "__hash__")
        return NotImplemented

class A:
    def __hash__(self) -> int:
        return 0

a: Hashable = A()
```

outputs

```
$ mypy demo.py --strict
demo.py:32: error: Incompatible types in assignment (expression has type "A", variable has type "Hashable")
Found 1 error in 1 file (checked 1 source file)
```

**Comment by AlexWaygood** [comment]
@maggyero, it would be better to ask questions like this [here](https://github.com/python/typing/discussions), rather than leaving comments on issues that have been closed for nearly 6 years 🙂

**Comment by maggyero** [comment]
@AlexWaygood Thanks for the suggestion, done: https://github.com/python/typing/discussions/1127#discussion-3998470
